packages/transaction-controller/src/types.ts

```diff
@@ -689,3 +741,24 @@ export type DefaultGasEstimates = {
    */
   maxPriorityFeePerGas?: string;
 };
+
+export type SwapOptions = {
+  hasApproveTx?: boolean;
+  meta?: Partial<TransactionMeta>;
+};
+
+/**
+ * Describes the event types emitted by the transaction controller.
+ */
+export enum TransactionEvent {
```

Comment:
Rather than using an enum for these events, I've tried a new pattern in my `PendingTransactionTracker` PR where we extend the `EventEmitter` interface to insist that any events being emitted, and any listeners created, have to match specific event name and argument pairings.

https://github.com/MetaMask/core/blob/1dc1e026009da51af6406f59815520e1b58b6f44/packages/transaction-controller/src/helpers/PendingTransactionTracker.ts#L44

Reply:
This is really an amazing idea, I've also noted for myself that we may want to reiterate on the events because there are bunch of `txId:stage` events. Shall we have a look at this on different PR, maybe the one I made all about events?